cpp/src/traversal/opg/load_balance.cuh

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef LB_CUH
+#define LB_CUH
+
+#include "vertex_binning.cuh"
+#include "worker_kernels.cuh"
+#include <graph.hpp>
+
+namespace detail {
+
+namespace opg {
+
+template <typename VT, typename ET, typename WT>
+class LoadBalanceExecution {
+  raft::handle_t const &handle_;
+  cugraph::experimental::GraphCSRView<VT, ET, WT>& graph_;
+  VertexBinner<VT, ET> dist_;
+  rmm::device_vector<VT> reorganized_vertices_;
+  ET vertex_begin_;
+  ET vertex_end_;
+
+  public:
+  LoadBalanceExecution(raft::handle_t const & handle,
+      cugraph::experimental::GraphCSRView<VT, ET, WT>& graph) :
+    handle_(handle),
+    graph_(graph),
+    reorganized_vertices_(graph.local_vertices[handle_.get_comms().get_rank()]),
+    vertex_begin_(graph.local_offsets[handle_.get_comms().get_rank()]),
+    vertex_end_(graph.local_offsets[handle_.get_comms().get_rank()] +
+        graph.local_vertices[handle_.get_comms().get_rank()])
+  {}
+
+  template <typename Operator>
+  void run(
+      Operator op,
+      unsigned *active_bitmap) {
+    cudaStream_t stream = handle_.get_stream();
+    dist_.setup(graph_.offsets, active_bitmap, vertex_begin_, vertex_end_);
+    auto distribution = dist_.run(reorganized_vertices_, stream);
+
+    DegreeBucket<VT, ET> large_bucket = distribution.degreeRange(16);
+    large_vertex_worker(graph_, large_bucket, op, stream);
+
+    DegreeBucket<VT, ET> medium_bucket = distribution.degreeRange(12, 16);
+    medium_vertex_worker(graph_, medium_bucket, op, stream);
+
+    DegreeBucket<VT, ET> small_bucket_0 = distribution.degreeRange(10,12);
```

Comment:
The choice to map kernels to the degree ranges has more to do with the performance of the kernel rather than characteristics of the graph. We can do some more work to check out what kernel is better suited to a degree later on.